docs/apis/savepoints.md

```diff
@@ -0,0 +1,108 @@
+
+### JobManager
+
+This is the **default backend** for savepoints.
+
+Savepoints are stored on the heap of the job manager. They are *lost* after the job manager is shut down. This mode is only useful if you want to *stop* and *resume* your program while the **same cluster** keeps running. It is *not recommended* for production use. Savepoints are *not* part the [job manager's highly availabile]({{ site.baseurl }}/setup/jobmanager_high_availability.html) state.
+
+<pre>
+state.backend.savepoints: jobmanager
+</pre>
+
+**Note**: If you don't configure a specific state backend for the savepoints, the default state backend (config key `state.backend`) will be used.
+
+### File system
+
+Savepoints are stored in the configured **file system directory**. They are available between cluster instances and allow to move your program to another cluster.
+
+<pre>
+state.backend.savepoints: filesystem
+state.backend.savepoints.fs.dir: hdfs:///flink/savepoints
+</pre>
+
+**Note**: If you don't configure a specific directory, the checkpoint directory (config key `state.backend.fs.checkpointdir`) will be used.
+
+**Important**: A savepoint is a pointer to completed checkpoint. That means that the state of a savepoint is not only found in the savepoint file itself, but also needs the actual checkpoint data (e.g. in a set of further files).
+
+## Changes to your program
+
+Savepoints **work out of the box**, but it is **highly recommended** that you slightly adjust your programs in order to be able to work with savepoints in future versions of your program.
+
+<img src="fig/savepoints-program_ids.png" class="center" />'
+
+For savepoints **only stateful tasks matter**. In the above example, the source and map tasks are stateful whereas the sink is not stateful. Therefore, only the state of the source and map tasks are part of the savepoint.
+
+Each task is identified by its **generated task IDs** and **subtask index**. In the above example the state of the source (**s<sub>1</sub>**, **s<sub>2</sub>**) and map tasks (**m<sub>1</sub>**, **m<sub>2</sub>**) is identified by their respective task ID (*0xC322EC* for the source tasks and *0x27B3EF* for the map tasks) and subtask index. There is no state for the sinks (**t<sub>1</sub>**, **t<sub>2</sub>**). Their IDs therefore do not matter.
+
+The IDs are generated **deterministically** from your program structure. This means that if your program does not change, the IDs do not change. In this case, it is straight forward to restore the state from a savepoint by mapping it back to the same task IDs and subtask indexes. This allows you to work with savepoints out of the box, but gets problematic as soon as you make changes to your program, because they result in changed IDs and the savepoint state cannot be mapped to your program any more.
```

Comment:
But changes to the UDF are allowed, right? If so, then we should state that here more clearly.